cpp/src/replace/replace.cu

```diff
@@ -13,26 +16,551 @@
+/* --------------------------------------------------------------------------*/
+/**
+ * @brief Replace elements from `input_col` according to the mapping `values_to_replace` to
+ *        `replacement_values`, that is, replace all `values_to_replace[i]` present in `input_col`
+ *        with `replacement_values[i]`.
+ *
+ * @param[in] col gdf_column with the data to be modified
+ * @param[in] values_to_replace gdf_column with the old values to be replaced
+ * @param[in] replacement_values gdf_column with the new values
+ *
+ * @returns output gdf_column with the modified data
+ */
+/* ----------------------------------------------------------------------------*/
+  std::unique_ptr<cudf::column> find_and_replace_all(cudf::column_view const& input_col,
+                                                     cudf::column_view const& values_to_replace,
+                                                     cudf::column_view const& replacement_values,
+                                                     rmm::mr::device_memory_resource* mr){
+    return detail::find_and_replace_all(input_col, values_to_replace, replacement_values, mr, 0);
+  }
+} //end experimental
+} //end cudf
+
+namespace{ //anonymous
+
+template <typename Type>
+__global__
+void replace_nulls_with_scalar(cudf::size_type size,
+                               const Type* __restrict__ in_data,
+                               const cudf::bitmask_type* __restrict__ in_valid,
+                               const Type* __restrict__ replacement,
+                               Type* __restrict__ out_data)
+{
+  int tid = threadIdx.x;
+  int blkid = blockIdx.x;
+  int blksz = blockDim.x;
+  int gridsz = gridDim.x;
+
+  int start = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  for (int i=start; i<size; i+=step) {
+    out_data[i] = cudf::bit_is_set(in_valid, i)? in_data[i] : *replacement;
+  }
+}
+
+
+template <typename Type>
+__global__
+void replace_nulls_with_column(cudf::size_type size,
+                               Type const* __restrict__ in_data,
+                               cudf::bitmask_type const* __restrict__ in_valid,
+                               Type const* __restrict__ replacement,
+                               Type* __restrict__ out_data)
+{
+  int tid = threadIdx.x;
+  int blkid = blockIdx.x;
+  int blksz = blockDim.x;
+  int gridsz = gridDim.x;
+
+  int start = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  for (int i=start; i<size; i+=step) {
+    out_data[i] = cudf::bit_is_set(in_valid, i)? in_data[i] : replacement[i];
+  }
+}
+
+
+/* --------------------------------------------------------------------------*/
+/**
+ * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
+ *        `replace_nulls` with the appropriate data types.
+ */
+/* ----------------------------------------------------------------------------*/
+struct replace_nulls_column_kernel_forwarder {
+  template <typename col_type>
```

Comment:
This should use SFINAE to only allow fixed width types, i.e., `enable_if_t<is_fixed_width<col_type>()`.

Reply:
Done.